icu4c/source/i18n/formatted_string_builder.cpp

```diff
@@ -257,16 +258,37 @@ int32_t FormattedStringBuilder::prepareForInsert(int32_t index, int32_t count, U
     U_ASSERT(count >= 0);
     if (index == 0 && fZero - count >= 0) {
         // Append to start
+        int32_t newLength;
+        // Add and return U_INPUT_TOO_LONG_ERROR if overflow int32_t.
+        if (uprv_add32_overflow(fLength, count, &newLength)) {
+          status = U_INPUT_TOO_LONG_ERROR;
+          return count;
+        }
         fZero -= count;
-        fLength += count;
+        fLength = newLength;
         return fZero;
-    } else if (index == fLength && fZero + fLength + count < getCapacity()) {
-        // Append to end
-        fLength += count;
-        return fZero + fLength - count;
     } else {
+      int32_t position, requiredCapacity;
+      // Add and return U_INPUT_TOO_LONG_ERROR if overflow int32_t.
+      if (uprv_add32_overflow(fLength, fZero, &position) ||
```

Comment:
No need for a fancy check for `uprv_add32_overflow(fLength, fZero, &position)` because you would have caught an overflow there in a previous call. Just use `int32_t position = fZero + fLength;`

More importantly, see the comment on old lines 263..266 -- I think you can keep this whole case simple.

Reply:
how can we be sure fLength + fZero won't overflow here?

Reply:
done